Docs: Add environment variable reference page (#3755)
ROCmSoftwarePlatform/MIOpen: docs/reference/env_variables.rst

Reviewer: I think there is quite a few more specific solver environment variables not listed here.

For example these appear to be missing, and I imagine there might be more:
- MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_HIP_FWD_XDLOPS
- MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_HIP_BWD_XDLOPS
- MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_HIP_WRW_XDLOPS
- MIOPEN_DEBUG_3D_CONV_IMPLICIT_GEMM_HIP_FWD_XDLOPS
- MIOPEN_DEBUG_3D_CONV_IMPLICIT_GEMM_HIP_BWD_XDLOPS
- MIOPEN_DEBUG_3D_CONV_IMPLICIT_GEMM_HIP_WRW_XDLOPS
- MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_ASM_FWD_GTC_XDLOPS_NHWC
- MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_ASM_BWD_GTC_XDLOPS_NHWC
- MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_ASM_WRW_GTC_XDLOPS_NHWC

Author: Thanks, I'll look these up and add them in - it would help if someone from the team can also help identify any other missing environment variable, as currently the only means of checking is via github search

Author: I've added these so far (and a few extra), currently doing another pass through github to see if I can find more

Reviewer: Ill ask others to take a look. 
Thanks for updating!

.. meta::
  :description: MIOpen environment variables reference
  :keywords: MIOpen, ROCm, API, environment variables, environment, reference

********************************************************************
MIOpen environment variables
********************************************************************

This section describes the important MIOpen environment variables,
which are grouped by functionality.

> Logging and Debugging

Reviewer 2: ```suggestion
Logging and debugging
```

======================

The logging and debugging environment variables for MIOpen are collected in the following table.

> For more information, see :doc:`Log and debug <../how-to/debug-log>`.

Reviewer 2: I'm wondering if the link text s/b logging and debugging? The ToC is "log and debug" under the How-To, but the page is Logging and debugging.

Author: I think it makes better sense to use the title from the page, so I made the change to all links in the page

        | 4: All errors and warnings
        | 5: Info level debugging
        | 6: Detailed debugging information
        | 7: Trace level with additional details

    * - | ``MIOPEN_ENABLE_LOGGING_MPMT``
        | Prefixes each log line with process/thread identification for multi-process/multi-threaded debugging.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_ENABLE_LOGGING_ELAPSED_TIME``
        | Adds timestamp showing elapsed time in milliseconds since previous log message.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_CHECK_NUMERICS``
        | Scans inputs/outputs for numerical abnormalities (inf, NaN, zeros).
      - | 0x01: Fully informative (print all check results)
        | 0x02: Warning information (print only abnormalities)
        | 0x04: Throw error on detection
        | 0x08: Abort on abnormal result
        | 0x10: Print statistics (mean/absmean/min/max)

    * - | ``MIOPEN_DEBUG_CHECK_SUB_BUFFER_OOB_MEMORY_ACCESS``
        | Checks for GPU sub-buffer out-of-bounds memory access errors.
      - | 0 or unset: No OOB detection
        | 1: Check for OOBs before sub-buffer start
        | 2: Check for OOBs after sub-buffer end

> Find Mode Configuration

Reviewer 2: ```suggestion
Find mode configuration
```

and :doc:`Performance database <../conceptual/perfdb>`.

.. list-table::
    :header-rows: 1
    :widths: 70,30

    * - **Environment variable**
      - **Value**

    * - | ``MIOPEN_FIND_MODE``
        | Sets find mode to accelerate find API calls.
      - | "NORMAL" or 1: Full find mode (benchmarks all solvers)
        | "FAST" or 2: Fast find (use FindDb or immediate fallback)
        | "HYBRID" or 3: Hybrid find (FindDb hit or full find)
        | 4: Reserved (do not use)
        | "DYNAMIC_HYBRID" or 5: Dynamic hybrid (default, skip non-dynamic kernels)

    * - | ``MIOPEN_FIND_ENFORCE``
        | Controls auto-tune behavior and database updates.
      - | "NONE" or 1: No change in default behavior
        | "DB_UPDATE" or 2: Always perform auto-tune and update PerfDb
        | "SEARCH" or 3: Auto-tune even if not requested via API
        | "SEARCH_DB_UPDATE" or 4: Combination of DB_UPDATE and SEARCH
        | "DB_CLEAN" or 5: Remove optimized values from User PerfDb

    * - | ``MIOPEN_DEBUG_DISABLE_FIND_DB``
        | Disables FindDb functionality.
      - | 1: Disable FindDb
        | 0 or unset: Enable FindDb

> Algorithm Control

Reviewer 2: ```suggestion
Algorithm control
```

=================

The algorithm control environment variables for MIOpen are collected in the following table.
For more information, see :doc:`Log and debug <../how-to/debug-log>`.

.. list-table::
    :header-rows: 1
    :widths: 70,30

    * - **Environment variable**
      - **Value**

    * - | ``MIOPEN_DEBUG_CONV_FFT``
        | Controls FFT convolution algorithm.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_DIRECT``
        | Controls direct convolution algorithm.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_GEMM``
        | Controls GEMM convolution algorithm.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_WINOGRAD``
        | Controls Winograd convolution algorithm.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_IMPLICIT_GEMM``

>         | Controls Implicit GEMM convolution algorithm.

Reviewer 2: Should implicit be capitalized here?

Author: It shouldn't, I forgot about the rule for this and the other PR

      - | 0: Disable
        | 1: Enable

> Kernel Build Method Control

Reviewer 2: ```suggestion
Kernel build method control
```

===========================

The kernel build method control environment variables for MIOpen are collected in the following table.
For more information, see :doc:`Log and debug <../how-to/debug-log>`.

.. list-table::
    :header-rows: 1
    :widths: 70,30

    * - **Environment variable**
      - **Value**

    * - | ``MIOPEN_DEBUG_GCN_ASM_KERNELS``
        | Controls assembly language kernels for convolutions and batch normalization.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_HIP_KERNELS``
        | Controls HIP-written convolution kernels (ImplicitGemm algorithm).
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_OPENCL_CONVOLUTIONS``
        | Controls OpenCL-written convolution kernels.
      - | 0: Disable
        | 1: Enable

> Solution Selection

Reviewer 2: ```suggestion
Solution selection
```

==================

The solution selection environment variables for MIOpen are collected in the following table.
For more information, see :doc:`Log and debug <../how-to/debug-log>`.

.. list-table::
    :header-rows: 1
    :widths: 70,30

    * - **Environment variable**
      - **Value**

    * - | ``MIOPEN_DEBUG_FIND_ONLY_SOLVER``

>         | Forces use of only one specific solution. Affects Find() calls only.

Reviewer 2: If Find() is a function, place it in code ticks.

      - | Numeric or string solution identifier
        | If valid and applicable: only that solution is found
        | If valid but not applicable: Find() fails
        | If invalid: Find() call fails

> Direct Solution Control

Reviewer 2: ```suggestion
Direct solution control
```

        | Controls ConvAsmBwdWrW1x1 direct solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_DIRECT_OCL_FWD11X11``
        | Controls ConvOclDirectFwd11x11 direct solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_DIRECT_OCL_FWDGEN``
        | Controls ConvOclDirectFwdGen direct solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_DIRECT_OCL_FWD``
        | Controls ConvOclDirectFwd direct solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_DIRECT_OCL_FWD1X1``
        | Controls ConvOclDirectFwd1x1 direct solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_DIRECT_OCL_WRW2``
        | Controls ConvOclBwdWrW2<n> (n={1,2,4,8,16}) and ConvOclBwdWrW2NonTunable solutions.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_DIRECT_OCL_WRW53``
        | Controls ConvOclBwdWrW53 direct solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_DIRECT_OCL_WRW1X1``
        | Controls ConvOclBwdWrW1x1 direct solution.
      - | 0: Disable
        | 1: Enable

> Winograd Solution Control

Reviewer 2: ```suggestion
Winograd solution control
```

=========================

The Winograd solution control environment variables for MIOpen are collected in the following table.
For more information, see :doc:`Log and debug <../how-to/debug-log>`.

.. list-table::
    :header-rows: 1
    :widths: 70,30

    * - **Environment variable**
      - **Value**

    * - | ``MIOPEN_DEBUG_AMD_WINOGRAD_3X3``
        | Controls ConvBinWinograd3x3U FP32 Winograd Fwd/Bwd (filter size 3x3).
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_AMD_WINOGRAD_RXS``
        | Controls ConvBinWinogradRxS FP32/FP16 F(3,3) Fwd/Bwd and FP32 F(3,2) WrW Winograd.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_AMD_WINOGRAD_RXS_WRW``

>         | Controls FP32 F(3,2) WrW convolutions only (subset of MIOPEN_DEBUG_AMD_WINOGRAD_RXS).

Reviewer 2: ```suggestion
        | Controls FP32 F(3,2) WrW convolutions only (subset of ``MIOPEN_DEBUG_AMD_WINOGRAD_RXS``).
```

      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_AMD_WINOGRAD_RXS_FWD_BWD``

>         | Controls FP32/FP16 F(3,3) Fwd/Bwd (subset of MIOPEN_DEBUG_AMD_WINOGRAD_RXS).

Reviewer 2: ```suggestion
        | Controls FP32/FP16 F(3,3) Fwd/Bwd (subset of ``MIOPEN_DEBUG_AMD_WINOGRAD_RXS``).
```

      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_AMD_WINOGRAD_RXS_F3X2``
        | Controls ConvBinWinogradRxSf3x2 FP32/FP16 Fwd/Bwd F(3,2) Winograd.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_AMD_WINOGRAD_RXS_F2X3``
        | Controls ConvBinWinogradRxSf2x3 FP32/FP16 Fwd/Bwd F(2,3) Winograd (group convolutions only).
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_AMD_WINOGRAD_RXS_F2X3_G1``
        | Controls ConvBinWinogradRxSf2x3g1 FP32/FP16 Fwd/Bwd F(2,3) Winograd (non-group convolutions).
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_AMD_FUSED_WINOGRAD``
        | Controls Fused FP32 F(3,3) Winograd with variable filter size.
      - | 0: Disable
        | 1: Enable

> Multi-pass Winograd Solution Control

Reviewer 2: ```suggestion
Multi-pass Winograd solution control
```

        | Controls ConvMPBidirectWinograd_xdlops<5-3> FWD/BWD F(5,3).
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_AMD_MP_BD_XDLOPS_WINOGRAD_F6X3``
        | Controls ConvMPBidirectWinograd_xdlops<6-3> FWD/BWD F(6,3).
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_AMD_MP_BD_WINOGRAD_EXPEREMENTAL_FP16_TRANSFORM``
        | Controls ConvMPBidirectWinograd* FWD/BWD FP16 experimental mode (use at your own risk).
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_AMD_WINOGRAD_MPASS_WORKSPACE_MAX``
        | Sets workspace size limit for ConvWinograd3x3MultipassWrW solutions.
      - | Decimal or hex value (64-bit unsigned integer) in bytes
        | Default: 2000000000 (~1.862 GiB) for gfx900 and gfx906/60
        | 0: Use default limit
        | 1: Prohibit workspace use
        | -1: Remove default limit

    * - | ``MIOPEN_DEBUG_AMD_MP_BD_WINOGRAD_WORKSPACE_MAX``
        | Sets workspace size limit for ConvMPBidirectWinograd solutions.
      - | Decimal or hex value (64-bit unsigned integer) in bytes
        | 0: Use default limit
        | 1: Prohibit workspace use
        | -1: Remove default limit

> ASM Implicit GEMM Solution Control

Reviewer 2: ```suggestion
ASM implicit GEMM solution control
```

    * - | ``MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_ASM_FWD_V4R1``
        | Controls ConvAsmImplicitGemmV4R1DynamicFwd solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_ASM_FWD_V4R1_1X1``
        | Controls ConvAsmImplicitGemmV4R1DynamicFwd_1x1 solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_ASM_BWD_V4R1``
        | Controls ConvAsmImplicitGemmV4R1DynamicBwd solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_ASM_WRW_V4R1``
        | Controls ConvAsmImplicitGemmV4R1DynamicWrw solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_ASM_FWD_GTC_XDLOPS``
        | Controls ConvAsmImplicitGemmGTCDynamicFwdXdlops solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_ASM_BWD_GTC_XDLOPS``
        | Controls ConvAsmImplicitGemmGTCDynamicBwdXdlops solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_ASM_WRW_GTC_XDLOPS``
        | Controls ConvAsmImplicitGemmGTCDynamicWrwXdlops solution.
      - | 0: Disable
        | 1: Enable

> HIP Implicit GEMM Solution Control

Reviewer 2: ```suggestion
HIP implicit GEMM solution control
```

        | Controls ConvHipImplicitGemmForwardV4R4Xdlops solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_HIP_FWD_V4R5_XDLOPS``
        | Controls ConvHipImplicitGemmForwardV4R5Xdlops solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_HIP_BWD_V1R1_XDLOPS``
        | Controls ConvHipImplicitGemmBwdDataV1R1Xdlops solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_HIP_BWD_V4R1_XDLOPS``
        | Controls ConvHipImplicitGemmBwdDataV4R1Xdlops solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_HIP_WRW_V4R4_XDLOPS``
        | Controls ConvHipImplicitGemmWrwV4R4Xdlops solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_HIP_FWD_V4R4_PADDED_GEMM_XDLOPS``
        | Controls ConvHipImplicitGemmForwardV4R4Xdlops_Padded_Gemm solution.
      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_HIP_WRW_V4R4_PADDED_GEMM_XDLOPS``
        | Controls ConvHipImplicitGemmWrwV4R4Xdlops_Padded_Gemm solution.
      - | 0: Disable
        | 1: Enable

> GEMM Backend Control

Reviewer 2: ```suggestion
GEMM backend control
```

====================

The GEMM backend control environment variables for MIOpen are collected in the following table.
For more information, see :doc:`Log and debug <../how-to/debug-log>`.

.. list-table::
    :header-rows: 1
    :widths: 70,30

    * - **Environment variable**
      - **Value**

    * - | ``MIOPEN_GEMM_ENFORCE_BACKEND``
        | Overrides default GEMM backend (rocBLAS).
      - | 1: Use rocBLAS if enabled
        | 2: Reserved
        | 3: No GEMM is called
        | 4: Reserved
        | 5: Use hipBLASLt if enabled
        | Any other value: Use default behavior

    * - | ``ROCBLAS_LAYER``
        | Controls rocBLAS GEMM logging output.
      - | Unset: No logging
        | 1: Trace logging
        | 2: Bench logging
        | 3: Trace and bench logging

    * - | ``HIPBLASLT_LOG_LEVEL``
        | Controls hipBLASLt GEMM logging output.
      - | 0: Off (default)
        | 1: Error logging
        | 2: Trace (API calls with parameters)
        | 3: Hints (performance improvement suggestions)
        | 4: Info (general execution information)

>         | 5: API Trace (detailed API parameters)

Reviewer 2: ```suggestion
        | 5: API trace (detailed API parameters)
```

> Convolution Attributes

Reviewer 2: ```suggestion
Convolution attributes
```

======================

The convolution attribute environment variables for MIOpen are collected in the following table.
For more information, see :doc:`MI200 alternate implementation <../conceptual/MI200-alt-implementation>`.

.. list-table::
    :header-rows: 1
    :widths: 70,30

    * - **Environment variable**
      - **Value**

    * - | ``MIOPEN_DEBUG_CONVOLUTION_ATTRIB_FP16_ALT_IMPL``

>         | Controls alternate FP16 implementation that uses BFloat16's larger exponent range for all convolution directions.

Reviewer 2: ```suggestion
        | Controls the alternate ``FP16`` implementation that uses the larger ``BFloat16`` exponent range for all convolution directions.
```

      - | 0: Disable
        | 1: Enable

    * - | ``MIOPEN_CONVOLUTION_ATTRIB_FP16_ALT_IMPL``

>         | Controls alternate FP16 implementation that uses BFloat16's larger exponent range (alternative to miopenSetConvolutionAttribute API).

Reviewer 2: ```suggestion
        | Controls the alternate ``FP16`` implementation that uses the larger ``BFloat16`` exponent range (alternative to the miopenSetConvolutionAttribute API).
```

      - | 0: Disable
        | 1: Enable

> Compilation Control

Reviewer 2: ```suggestion
Compilation control
```

===================

The compilation control environment variables for MIOpen are collected in the following table.
For more information, see :doc:`Log and debug <../how-to/debug-log>`.

.. list-table::
    :header-rows: 1
    :widths: 70,30

    * - **Environment variable**
      - **Value**

    * - | ``MIOPEN_COMPILE_PARALLEL_LEVEL``

>         | Controls parallel compilation thread count for Find() calls.

Reviewer 2: Check whether Find() s/b in code ticks

      - | Integer value (default: 20 threads)
        | 1: Disable multi-threaded compilation

> Experimental Controls

Reviewer 2: ```suggestion
Experimental controls
```

>       - | Integer value (default: 20 threads)

Reviewer: I think the default is 1 actually, but it's conditional:

```
#if MIOPEN_USE_COMGR
MIOPEN_DECLARE_ENV_VAR_UINT64(MIOPEN_COMPILE_PARALLEL_LEVEL, 1) // COMGR is not parallelizable
#else
MIOPEN_DECLARE_ENV_VAR_UINT64(MIOPEN_COMPILE_PARALLEL_LEVEL,
                              std::thread::hardware_concurrency() / 2)
#endif
```

Author: Fixed

=====================

The experimental control environment variables for MIOpen are collected in the following table.
For more information, see :doc:`Log and debug <../how-to/debug-log>`.

.. list-table::
    :header-rows: 1
    :widths: 70,30

    * - **Environment variable**
      - **Value**

    * - | ``MIOPEN_DEBUG_AMD_ROCM_METADATA_ENFORCE``

>         | Controls Code Object version for GCN assembly kernels.

Reviewer 2: ```suggestion
        | Controls Code Object (CO) version for GCN assembly kernels.
```
